Documentation: make a gitbook for documentation

curioussavage opened the pull request:
:eyeglasses: @piranna 

This commit addresses https://github.com/blessed-ng/blessed/issues/1

This commit does not make any changes to the documentation itself. It only moves it
into the docs dir and separates it into files. There are also added files and npm scripts to handle
gitbook stuff. I followed what the [redux](https://github.com/reactjs/redux) repo does with their gitbook setup. It looks pretty nice to me. Let me know what you think.

I also purchased blessedjs.org so we can use that. I would need to add a CNAME file to this commit for the gh-pages branch deploy script to work. Not sure what other configuration is necessary (other than adding records to dns for the name).

curioussavage committed:
Documentation: make a gitbook for documentation

This moves the docs out of the main readme.md and
into a /docs dir. They were split out and made to
conform to the gitbook format.

piranna commented on .bookignore:
```diff
@@ -0,0 +1,9 @@
+bin/*
```
Move file to `/docs`

piranna commented on .gitignore:
```diff
@@ -1 +1,2 @@
 node_modules/
```
Move file to `/docs`

piranna commented on book.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Move file to `/docs`

piranna commented on docs/readme.md:
```diff
@@ -0,0 +1,83 @@
+# Widgets
+
+## base nodes
+* [Node](/docs/widgets/baseNodes/node.md) (abstract)
```
Use relative paths so pages can navigate on Github.

curioussavage replied:
it looks like github actually resolves these urls correctly. A few of the links are camel case where the file is all lower and that breaks it but it is navigable as is.

piranna replied:
> t looks like github actually resolves these urls correctly

Interesting... probably it resolves to the root of the repo...


> A few of the links are camel case where the file is all lower and that breaks it but it is navigable as is

I think they should be checked and fixed :-/

piranna commented on docs/widgets/baseNodes/element.md:
```diff
@@ -0,0 +1,201 @@
+#### Element (from Node)
```
Add hyperlink on the parent classes (they can be put on the title itself without problems)

curioussavage replied:
This would be a follow up commit.

I can put it in the pull request if you like in fact I think would be good so they are a little better to navigate ( although the gitbook has an always present navbar on the left. /shrug

piranna replied:
> This would be a follow up commit.

Yes, it could be added to this pull-request.

piranna commented on package.json:
```diff
@@ -6,6 +6,12 @@
   "license": "MIT",
   "main": "./lib/blessed.js",
   "bin": "./bin/tput.js",
+  "scripts": {
```
Set own `package.json` on `/docs` for the book itself, as if it were a subproject.

piranna commented on package.json:
```diff
@@ -6,6 +6,12 @@
+    "docs:build": "npm run docs:prepare && gitbook build",
+    "docs:watch": "npm run docs:prepare && gitbook serve",
+    "docs:publish": "rm -rf _book && npm run docs:build && cp CNAME _book && cd _book && git init && git commit --allow-empty -m 'update book' && git checkout -b gh-pages && touch .nojekyll && git add . && git commit -am 'update book' && git push git@github.com:blessed-ng/blessed gh-pages --force"
+  },
```
Are they needed all this steps? So far I only remember to need to define `gitbook install` as the `install` script of the book itself... :-/

curioussavage replied:
docs:watch of course is just for convenience.

piranna replied:
> docs:watch of course is just for convenience

I think `docs:publish` could be removed too, it's too big...

piranna commented:
> This commit does not make any changes to the documentation itself. It only moves it
> into the docs dir and separates it into files. There are also added files and npm scripts to handle
> gitbook stuff. I followed what the redux repo does with their gitbook setup. It looks pretty nice to me. Let me know what you think.

What such an awesome work, congrats! It's just what I was asking for :-) I have give you some comments about things I think needs to be improved. If you want you can take a look to my [bachelor thesis](https://github.com/piranna/pfc) as reference :-)


> I also purchased blessedjs.org so we can use that. I would need to add a CNAME file to this commit for the gh-pages branch deploy script to work. Not sure what other configuration is necessary (other than adding records to dns for the name).

I don't know why you need the CNAME, so far I only needed to do a transparent redirection to the Github Page url... Also, by using the `/docs` folder we don't need the `gh-pages` branch anymore :-)

curioussavage commented:
Sorry I think my explanation was not very comprehensive. The Redux repository I linked to has an example of how this would work if want to take a look. The issue with moving everything into the docs folder **and** wanting to serve from that folder is that you would then have to commit the `/_book` directory that gitbook outputs to the `/docs` dir. I'm not really a fan of that myself ( huge diffs for one). I think this setup they use is clever(in a good way) because the main readme of the repository becomes the introduction to the book. also you just run the script which commits to that branch and the site is served from main project directory.

piranna commented:
> you would then have to commit the /_book directory that gitbook outputs to the /docs dir

That's not needed, the point is that Github will render and publish directly from the Markdown files, not need to publish the Gitbook converted files. In fact, `_book` directory should be on the `.gitignore`.

curioussavage commented:
Another thing I didn't really say clearly - The solution currently implemented here avoids using the github jekyll site stuff in favor of using the site gitbook generates which gets a simple search and nice navigation for free.

piranna commented:
Ok... in that case I would like to use the Jekyll pages. We could use the
Gitbook search by publishing the docs there too.

El 17/2/2017 2:31, "curioussavage" <notifications@github.com> escribió:

Another thing I didn't really say clearly - The solution currently
implemented here avoids using the github jekyll site stuff in favor of
using the site gitbook generates which gets a simple search and nice
navigation for free.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/blessed-ng/blessed/pull/3#issuecomment-280522035>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAgfvvzuH1MeD7s2YuLCCIDxYLakVZUEks5rdPhWgaJpZM4MCqaR>
.

curioussavage commented:
I would rather go with one and forget about the other. Not much point in having two websites available. 

We could find use a plugin for search on the jekyll website.

piranna commented:
> I would rather go with one and forget about the other. Not much point in having two websites available.

Then let's go with Github pages, the Gitbook is just an extra.

curioussavage commented:
hey @piranna I haven't had much time to work on this lately but I want to get the site running so we can start improving the actual content of the docs. How can I convince you to go with the original solution? I think a gitbook site hosted on github is the nicest solution both for development and for users. 

Out of the box the sites look better and search works better. I know it adds a few extra files but if you check out the [jekyll project site itself](https://github.com/jekyll/jekyll/tree/master/docs) you will see that a nice jekyll site ends up with a bunch of extra files too. With the gitbook site on the other branch at least much of that isn't visible on master. Jekyll is made for blogs and it shows. Documentation sites seem like they are more of a first class feature of gitbook.

Another annoyance with the jekyll site is I couldn't get relative links to work at all. I noticed that the jekyll projects jekyll doc site itself uses absolute urls :(

I think you had an issue with the large npm scripts - we could easily move those to a file. If there are other changes that would make you happier let me know. I am under the assumption that I am going to be doing most of the work to improve the documentation in the near future and this would make it so much easier.

piranna commented:
> I think you had an issue with the large npm scripts

I don't have an issue with them, but with overcomplicate things and upload generated files, specially when they are not needed. Github allow to serve transparently a Github Pages site from a bunch of MarkDown files directly from the `docs` folder, so let's stick to them and just add there the Gitbook metadata files, no more. I don't think it's needed anymore... I could be able to fork your changes and do it myself if you don't understand what I'm talking about...

curioussavage commented:
> Github allow to serve transparently a Github Pages site from a bunch of MarkDown files directly from the docs folder

That's not exactly true. There is no such thing as a 'github pages' site. They support Jekyll which is a blog focused static site generator. It happens to support serving markdown files outside its opinionated folder structure with some configuration. It will not just convert any markdown files in the docs directory. See the jekyll projects on site in its docs folder for reference. 

Once you add plugins to get features like search which good docs *should* have and you add a nice theme (all the built in github themes are very bare bones) It won't look so simple. Github's own documentation says its themes are only meant as starting points.

The only generated files in the `_book` directory are the html files. It's better than letting github build in an environment you don't control and sending you build errors with zero information. The rest of the files are static assets which we would have a lot of after making a jekyll site not suck. The nice thing about using a different branch is that doesn't clutter up the main repo at all. 

Another point - with gitbook you get relative urls that work. Jekyll doesn't handle relative URLs well so unless you want to use a hack like [this](http://ricostacruz.com/til/relative-paths-in-jekyll) guy - you lose the ability to navigate the docs in the repo.

Going the route the redux project did is, in my opinion, the _simplest_ way to get a *nice* documentation site (one that looks good and has good search).

edit: if you want to compare them then run the gitbook build command in this pull request and compare it to https://curioussavage.github.io/blessed/

curioussavage commented:
late comment but.

> I don't know why you need the CNAME, so far I only needed to do a transparent redirection to the Github Page url...

This is necessary with a custom domain whatever we end up going with. If you add a custom domain with githubs GUI all it does is add that file.

piranna commented:
> It will not just convert any markdown files in the docs directory

Yes, [it does](https://github.com/blog/2228-simpler-github-pages-publishing):

> Selecting master branch /docs folder will publish from the /docs folder of your master branch. This lets you maintain documentation and code together on one branch, and open source maintainers can accept contributions for both in a single pull request.


> Once you add plugins to get features like search which good docs should have

Having Google, nobody use a domain ad-hoc search engine... Maybe just only for offline docs. _Maybe_.


> Another point - with gitbook you get relative urls that work. Jekyll doesn't handle relative URLs well so unless you want to use a hack like this guy - you lose the ability to navigate the docs in the repo.

Why do you insists on going down the full path of using Jekyll? Are you trying to generate the content on your machine and just upload the static assets instead of let Github Pages do it for yourself?


> Going the route the redux project did is, in my opinion, the simplest way to get a nice documentation site (one that looks good and has good search).

I think we have diferent opinions about what's simplest here... For me, it's just to have them as simple markdown files, no more. Any other processing is superfluous.

piranna commented:
> I don't know why you need the CNAME, so far I only needed to do a transparent redirection to the Github Page url...
> This is necessary with a custom domain whatever we end up going with. If you add a custom domain with githubs GUI all it does is add that file.

As I told you before, I have just only needed to configure the redirection, no more.

curioussavage commented:
You are right about the markdown files. I was convinced it didn't do that because I forgot to remove the .md extension on my test site and it includes them in the compiled site so I was visiting original md files. Ill fix those URLS and it should work.

> Are you trying to generate the content on your machine and just upload the static assets

Of course not. But that would be nice in some ways. Like I said github has build errors and doesn't give you any details most of the time. 

I disagree about search, many good documentation sites have search. Its a waste time now though so I just force pushed a rebased fork of this pull request with fixes. Links are good now and Its a useable albeit crappy jekyll site.

 I want to make another pass and clean up some of the pages. There are a lot of files in widgets that just say that a widget inherits everything from other widgets. I might aggregate some of the folders of widgets into single files. Ill add links back to the element widget that everything inherits from too.

piranna commented:
You are right about the markdown files. I was convinced it didn't do that
because I forgot to remove the .md extension on my test site and it
includes them in the compiled site so I was visiting original md files. Ill
fix those URLS and it should work.

This is already done on the pull-request I've already send you. In fact, it
has automatically changed the URL extensions, the trick was to use relative
paths instead of absolute ones.


Are you trying to generate the content on your machine and just upload the
static assets

Of course not. But that would be nice in some ways. Like I said github has
build errors and doesn't give you any details most of the time.

So far I didn't found any of the errors you talk about... Could you be able
to specify them?

I want to make another pass and clean up some of the pages. There are a lot
of files in widgets that just say that a widget inherits everything from
other widgets. I might aggregate some of the folders of widgets into single
files. Ill add links back to the element widget that everything inherits
from too.

It would be just enough to add links to the parents classes, currently they
are set as local anchors. Some index.md files for the widgets and other
groups would be a good idea too.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/blessed-ng/blessed/pull/3#issuecomment-287202043>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAgfvuxHKbLiebZ5wlWFqPQkmCFSPHk2ks5rma7YgaJpZM4MCqaR>
.

curioussavage commented:
I did not see that pull request on my fork. Ill take a look at it tomorrow.

curioussavage commented:
@piranna. ok I had already rebased my own progress on the branch so It was easiest to just checkout your branch and force push it. I rebased it down to one commit too.

How about we call this good and I will start a new branch to work on some more improvements to the docs like more links/ examples / images etc.

curioussavage commented:
ok I added a DNS record with the github pages IP addresses. you should be able to add blessedjs.org as a custom domain.

piranna commented:
> How about we call this good and I will start a new branch to work on some more improvements to the docs like more links/ examples / images etc.

Agree. Done.


> ok I added a DNS record with the github pages IP addresses. you should be able to add blessedjs.org as a custom domain.

Not sure if It's possible, I'll try it.

piranna commented:
Ok... so it worked :-) Site is published :-)

curioussavage commented:
sweet!
